plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+    }
+
+    public void rebuild(final StorageTx tx)
+    {
+      final ODatabaseDocumentTx readDatabase = tx.getDb();
+      readDatabase.command(
+          new OSQLAsynchQuery<ODocument>(
+              sql,
+              new OCommandResultListener()
+              {
+                String currentGroupId = null;
+
+                @Override
+                public boolean result(Object iRecord) {
+                  try {
+                    final ODocument doc = (ODocument) iRecord;
+                    final String groupId = doc.field("groupId", OType.STRING);
+                    final String artifactId = doc.field("artifactId", OType.STRING);
+                    final Set<String> baseVersions = doc.field("baseVersions", OType.EMBEDDEDSET);
+
+                    final boolean groupChange = !Objects.equals(currentGroupId, groupId);
+                    if (groupChange) {
+                      if (currentGroupId != null) {
+                        processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                      }
+                      currentGroupId = groupId;
+                      metadataBuilder.onEnterGroupId(groupId);
+                    }
+                    rebuildAbV((ORID) sqlParams.get("bucket"), groupId, artifactId, baseVersions);
+                    return true;
+                  }
+                  finally {
+                    ODatabaseRecordThreadLocal.INSTANCE.set(readDatabase); // always reset thread DB
+                  }
+                }
+
+                @Override
+                public void end() {
+                  if (currentGroupId != null) {
+                    processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                  }
+                }
+              }
+          )
+      ).execute(sqlParams);
+    }
+
+    /**
+     * Method processing in separate TX/DB, performs writes (and mangles ThreadLocal DB, so caller should restore it).
+     * Handles A and bV on metadataBuilder.
+     */
+    private void rebuildAbV(final ORID bucket,
+                            final String groupId,
+                            final String artifactId,
+                            final Set<String> baseVersions)
+    {
+      metadataBuilder.onEnterArtifactId(artifactId);
+      for (String baseVersion : baseVersions) {
+        metadataBuilder.onEnterBaseVersion(baseVersion);
+        try (StorageTx tx = storageFacet.openTx()) {
+          final Iterable<Component> components = tx.findComponents(
+              "bucket = :bucket and group = :groupId and name = :artifactId and attributes.maven2.baseVersion = :baseVersion",
+              ImmutableMap
+                  .<String, Object>of("bucket", bucket, "groupId", groupId, "artifactId", artifactId, "baseVersion",
+                      baseVersion),
+              null,
+              "order by version asc");
```

Comment:
Basically, the "order by" is unimportant here, so will remove it: the `metadataBuilder` keeps internal state correct (preserves "max"), so ordering might be done as part of some optimisation maybe, later. But at the end, since Maven3 introduction of new format, you _must_ parse all artifacts there, as you have no guarantee that you will not find an older CE (as it was with Maven2).